In [122]:
#This code is developed by Pooyan :D

# Assignment Tittle
__Author__ : Pooyan oskouhi 400222003
__Course__ : Undergraduate Numerical Analysis Course

In [123]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Numpy Problems :

# Part a:
Shape of file in numpy.

# Solution
Use function __shape__.

In [124]:
df = pd.read_csv('tempratures.csv')
arr = np.array(df)
print(arr.shape)

(51, 3)


In [125]:
arr

array([['2022-01-01', 'New York', 0.5],
       ['2022-01-01', 'Los Angeles', 20.5],
       ['2022-01-01', 'Miami', 24.7],
       ['2022-01-02', 'New York', 2.3],
       ['2022-01-02', 'Los Angeles', 18.4],
       ['2022-01-02', 'Miami', 22.9],
       ['2022-01-03', 'New York', 4.8],
       ['2022-01-03', 'Los Angeles', 15.9],
       ['2022-01-03', 'Miami', 21.2],
       ['2022-01-04', 'New York', 3.2],
       ['2022-01-04', 'Los Angeles', 16.1],
       ['2022-01-04', 'Miami', 23.5],
       ['2022-01-05', 'New York', -1.9],
       ['2022-01-05', 'Los Angeles', 22.8],
       ['2022-01-05', 'Miami', 20.1],
       ['2022-01-06', 'New York', -2.1],
       ['2022-01-06', 'Los Angeles', 23.5],
       ['2022-01-06', 'Miami', 25.3],
       ['2022-01-07', 'New York', 1.3],
       ['2022-01-07', 'Los Angeles', 20.3],
       ['2022-01-07', 'Miami', 21.8],
       ['2022-01-08', 'New York', 4.2],
       ['2022-01-08', 'Los Angeles', 19.7],
       ['2022-01-08', 'Miami', 22.4],
       ['2022-01-09', 

# Part b:
Find the average temperature for each city.

# Solution
We sum all the elements that relate to a city and divide to number of days.

In [126]:
sumNY = 0
for i in range(0, 51, 3):
    sumNY += arr[i][2]
sumNY/12

0.225

In [127]:
sumLA = 0
for i in range(1, 51, 3):
    sumLA += arr[i][2]
sumLA/12

29.08333333333334

In [128]:
sumMiami = 0
for i in range(2, 52, 3):
    sumMiami += arr[i][2]
sumMiami/12

32.041666666666664

# Part c:
Find maximum temperature for each city.

# Solution
For each city find the maximum.

In [129]:
maxNY = 0
for i in range(0, 51, 3):
    if maxNY < arr[i][2]:
        maxNY = arr[i][2]
maxNY

5.8

In [130]:
maxLA = 0
for i in range(1, 51, 3):
    if maxLA < arr[i][2]:
        maxLA = arr[i][2]
maxLA

24.4

In [131]:
maxMiami = 0
for i in range(2, 52, 3):
    if maxMiami < arr[i][2]:
        maxMiami = arr[i][2]
maxMiami

25.8

# Pandas Problems :

# Part a:
Shape of file in pandas

# Solution
We use function __read_csv__ to read the data and after that use __shape__ to show th shape.

In [132]:
df = pd.read_csv('tempratures.csv', dtype=object)
df

date         city temperature
0   2022-01-01     New York         0.5
1   2022-01-01  Los Angeles        20.5
2   2022-01-01        Miami        24.7
3   2022-01-02     New York         2.3
4   2022-01-02  Los Angeles        18.4
5   2022-01-02        Miami        22.9
6   2022-01-03     New York         4.8
7   2022-01-03  Los Angeles        15.9
8   2022-01-03        Miami        21.2
9   2022-01-04     New York         3.2
10  2022-01-04  Los Angeles        16.1
11  2022-01-04        Miami        23.5
12  2022-01-05     New York        -1.9
13  2022-01-05  Los Angeles        22.8
14  2022-01-05        Miami        20.1
15  2022-01-06     New York        -2.1
16  2022-01-06  Los Angeles        23.5
17  2022-01-06        Miami        25.3
18  2022-01-07     New York         1.3
19  2022-01-07  Los Angeles        20.3
20  2022-01-07        Miami        21.8
21  2022-01-08     New York         4.2
22  2022-01-08  Los Angeles        19.7
23  2022-01-08        Miami        22.4
24  2022-01-09     New York         5.8
25  2022-01-09  Los Angeles        18.8
26  2022-01-09        Miami        20.5
27  2022-01-10     New York        -1.1
28  2022-01-10  Los Angeles        22.1
29  2022-01-10        Miami        23.1
30  2022-01-11     New York        -4.6
31  2022-01-11  Los Angeles        21.2
32  2022-01-11        Miami        25.8
33  2022-01-12     New York        -7.3
34  2022-01-12  Los Angeles        24.4
35  2022-01-12        Miami        22.6
36  2022-01-13     New York        -2.4
37  2022-01-13  Los Angeles        20.1
38  2022-01-13        Miami        19.7
39  2022-01-14     New York         1.5
40  2022-01-14  Los Angeles        21.5
41  2022-01-14        Miami        20.2
42  2022-01-15     New York         2.1
43  2022-01-15  Los Angeles        18.6
44  2022-01-15        Miami        21.5
45  2022-01-16     New York        -0.9
46  2022-01-16  Los Angeles        23.8
47  2022-01-16        Miami        23.9
48  2022-01-17     New York        -2.7
49  2022-01-17  Los Angeles        21.3
50  2022-01-17        Miami        25.3

In [133]:
df.shape

(51, 3)

In [134]:
df.describe()

date      city temperature
count           51        51          51
unique          17         3          44
top     2022-01-01  New York        20.1
freq             3        17           2

# Part b:
Calculate the median temperature for each city.

# Solution
We calculate medians through function __median()__.
__median() :__ This function calculate the median of values.

In [135]:
dfNY = df.loc[df['city'] == 'New York']
dfLA = df.loc[df['city'] == 'Los Angeles']
dfMiami = df.loc[df['city'] == 'Miami']

In [136]:
print(dfNY.median())
print(dfLA.median())
print(dfMiami.median())

temperature    0.5
dtype: float64
temperature    20.5
dtype: float64
temperature    22.6
dtype: float64


C:\Users\Lenovo\AppData\Local\Temp\ipykernel_7128\259389216.py:1: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  print(dfNY.median())
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_7128\259389216.py:2: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  print(dfLA.median())
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_7128\259389216.py:3: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select onl

# Part c:
Show the range of temperature in NY and LA and Miami.
Also, the maximum and minimum temperature of each city.

# Solution
We create for each city a data frame and analysis on that.

In [137]:
dfNY = df.loc[df['city'] == 'New York']
dfNY

date      city temperature
0   2022-01-01  New York         0.5
3   2022-01-02  New York         2.3
6   2022-01-03  New York         4.8
9   2022-01-04  New York         3.2
12  2022-01-05  New York        -1.9
15  2022-01-06  New York        -2.1
18  2022-01-07  New York         1.3
21  2022-01-08  New York         4.2
24  2022-01-09  New York         5.8
27  2022-01-10  New York        -1.1
30  2022-01-11  New York        -4.6
33  2022-01-12  New York        -7.3
36  2022-01-13  New York        -2.4
39  2022-01-14  New York         1.5
42  2022-01-15  New York         2.1
45  2022-01-16  New York        -0.9
48  2022-01-17  New York        -2.7

In [138]:
dfNY.plot(x='date', y='temperature')

TypeError: no numeric data to plot

In [ ]:
dfNY.min()

In [ ]:
dfNY.max()

So the range of the temperature is between -7.3 and 5.8 in New York.
The difference is 13.1.

In [ ]:
dfLA = df.loc[df['city'] == 'Los Angeles']
dfLA

In [ ]:
dfLA.plot(x='date', y='temperature')

In [ ]:
dfLA.max()

In [ ]:
dfLA.min()

So the range of the temperature is between 15.9 and 24.4 in Los Angeles.
The difference is 8.5.

In [ ]:
dfMiami = df.loc[df['city'] == 'Miami']
dfMiami

In [ ]:
dfMiami.plot(x='date', y='temperature')

In [ ]:
dfMiami.max()

In [ ]:
dfMiami.min()

So the range of the temperature is between 19.7 and 25.8 in Miami.
The difference is 6.1.